glib/src/value.rs

```diff
@@ -614,253 +767,195 @@ pub trait ToSendValue: Send + ToValue {
-    fn to_send_value(&self) -> SendValue {
-        SendValue(self.to_value())
-    }
-}
-
-impl<T: ?Sized + SetValue + Send + ToValue> ToSendValue for T {
+impl<T: Send + ToValue + ?Sized> ToSendValue for T {
     fn to_send_value(&self) -> SendValue {
         SendValue(self.to_value())
     }
 }
 
-/// Extracts a value.
-///
-/// Types that don't support a `None` value always return `Some`.
-pub trait FromValueOptional<'a>: StaticType + Sized {
-    /// # Safety
-    ///
-    /// The caller is responsible for ensuring the given `Value` is of a suitable
-    /// type for this conversion.
-    unsafe fn from_value_optional(value: &'a Value) -> Option<Self>;
-}
-
-/// Extracts a value.
-///
-/// Only implemented for types that don't support a `None` value.
-pub trait FromValue<'a>: FromValueOptional<'a> {
-    /// # Safety
-    ///
-    /// The caller is responsible for ensuring the given `Value` is of a suitable
-    /// type for this conversion.
-    unsafe fn from_value(value: &'a Value) -> Self;
-}
+unsafe impl<'a> FromValue<'a> for &'a str {
+    type Error = ValueTypeMismatchOrNoneError<Self>;
 
-/// Sets a value.
-///
-/// Only implemented for types that support a `None` value.
-pub trait SetValueOptional: SetValue {
-    /// # Safety
-    ///
-    /// The caller is responsible for ensuring the given `Value` is of a suitable
-    /// type for this conversion.
-    unsafe fn set_value_optional(value: &mut Value, new_value: Option<&Self>);
+    unsafe fn from_value(value: &'a Value) -> Self {
+        let ptr = gobject_ffi::g_value_get_string(value.to_glib_none().0);
+        CStr::from_ptr(ptr).to_str().expect("Invalid UTF-8")
+    }
 }
 
-/// Sets a value.
-pub trait SetValue: StaticType {
-    /// # Safety
-    ///
-    /// The caller is responsible for ensuring the given `Value` is of a suitable
-    /// type for this conversion.
-    unsafe fn set_value(value: &mut Value, new_value: &Self);
-}
+impl ToValue for str {
+    fn to_value(&self) -> Value {
+        unsafe {
+            let mut value = Value::from_type(<String>::static_type());
 
-impl<'a> FromValueOptional<'a> for String {
-    unsafe fn from_value_optional(value: &'a Value) -> Option<Self> {
-        from_glib_none(gobject_ffi::g_value_get_string(value.to_glib_none().0))
-    }
-}
+            gobject_ffi::g_value_take_string(value.to_glib_none_mut().0, self.to_glib_full());
 
-impl<'a> FromValueOptional<'a> for &'a str {
-    unsafe fn from_value_optional(value: &'a Value) -> Option<Self> {
-        let cstr = gobject_ffi::g_value_get_string(value.to_glib_none().0);
-        if cstr.is_null() {
-            None
-        } else {
-            CStr::from_ptr(cstr).to_str().ok()
+            value
         }
     }
-}
 
-impl SetValue for str {
-    unsafe fn set_value(value: &mut Value, this: &Self) {
-        gobject_ffi::g_value_take_string(value.to_glib_none_mut().0, this.to_glib_full())
+    fn value_type(&self) -> Type {
+        String::static_type()
     }
 }
 
```

Comment:
There can also be a `impl ValueType for &'static str { ... }` now, which can be useful.

Reply:
And should maybe also add impls for `Cow<str>`